Adding support for optional ContextKeys

bauer-andreas opened the pull request:
# Adding support for optional ContextKeys

## :recycle: Current situation

Defining `ContextKeys` with an optional type doesn't work properly (see #75).

## :bulb: Proposed solution

This PR adds a new `OptionalContextKey` protocol, which can be used for `ContextKey`s without a default value.
To reflect this behavior, a specific `Context.get(...)` was added to return an optional, to support the case where no value was provided for a given value.

### Problem that is solved

#75 

Some of the gRPC modifiers already needed that behavior and were adjusted.

I provided a default implementation for the `reduce(...)` function, which is widely used.

### Implications

Any `ContextKey` MUST NOT be defined with the associated type being a `Optional<...>`. I added documentation stating this and additionally a runtime check, however I couldn't figure out a way to move this to a compiler warning (not sure if this is possible).

I have tried ways of adding optional `ContextKeys` by defining the `Value` associated type as `Optional<...>`, however this was a bit unclean approach, as one need to still define a `defaultValue` and the signature of the `reduce(...)` function wasn't really fitting.

## :heavy_plus_sign: Additional Information

### Related PRs
* #117 The `DescriptionModifier` can make use of the new `OptionalContextKey` protocol.

### Testing
Test cases checking `OptionalContextKey`s were added.

### Reviewer Nudging
--

bauer-andreas committed:
Adding OptionalContextKey protocol to support optional ContextKeys

lukaskollmer commented:
Hmmm not really a fan of the duplication tbh. I fear that this would make it, when changing anything about this code in the future, way too easy to miss that the change in fact needs to be applied to two places.

Also not sure about the "lets just wrap everything in an optional" approach, my guess would be that the vast majority of `ContextKey`s are non-optional, so this would introduce the optional-caused additional overhead for all context values, even though only very few actually need it.

Maybe we can do a protocol-based approach to detect if a `ContextKey` has an optional value? A conditional extension on the `ContextKey` type or something like that?

Apart from all this, I think having an `OptionalContextKey` would probably be the cleanest solution. (Or would that also require duplication?)

bauer-andreas commented:
> Hmmm not really a fan of the duplication tbh. I fear that this would make it, when changing anything about this code in the future, way too easy to miss that the change in fact needs to be applied to two places.

Feel the same. And obviously the end goal is to remove all this code duplication.
However, before doing unnecessary work I wanted to discuss how we should implement options `ContextKey`s from an API point of view.

This PR started with just using the same `ContextKey` protocol by just declaring the associated type as an Optional, as having just one unified protocol would be a great thing I though. However this may create unnecessary complications when implementing the `reduce` function.

But yes, you are seemingly feeling the same that a dedicated protocol might yield a cleaner solution.

> Maybe we can do a protocol-based approach to detect if a ContextKey has an optional value? A conditional extension on the ContextKey type or something like that?

Detecting if the ContextKey holds an optional isn't a problem. I think you talk about a similar approach I used for extracting the type information from optional `Parameter`s right? Was also my first attempt, but was a nightmare to handle the types. 

I will got back to the drawing table and see what I can do.

bauer-andreas committed:
Adding documentation

bauer-andreas committed:
reduce wouldn't be called on the defaultValue.
Restored functionality and added a regression test for it

bauer-andreas commented:
Alright, so I have worked on an approach using a separate protocol and polished that one. I have updated the original message to reflect those changes, so maybe just have a look at it.

EDIT: also added some more regression test, so anyone in the future dealing with ContextNodes should be perfectly save making changes 👍

bauer-andreas committed:
Adding test case to test that ContextKex with optional types are rejected

moritzschuell commented on Sources/Apodini/SyntaxTreeVisitor/ContextKey.swift:
```diff
@@ -5,11 +5,46 @@
+}
+
+extension ContextKey {
+    // I know the compiler creates a warning about this, but using it as a type constraint isn't actually the same thing
```
Why is that the case? 🙃

bauer-andreas replied:
I would categorize this as "swift compiler weirdness"

If `ContextKey` was just defined like the following ...

```swift
protocol ContextKey: OptionalContextKey {
    static var defaultValue: Value { get }
}
```

... one would assume a definition like this is valid:

```swift
struct ExampleContextKey: ContextKey {
  static let defaultValue: String = "asdf"
}
```

It isn't. Although default value is defined to be type `String`, the swift compiler can't infer the type of the `associatedtype Value`  located in the parent protocol `OptionalContextKey` (I assume because the definition of the associated type isn't in the same protocol?).

My workaround then was to define the associated type `DefaultValue` and just assign it to `Value`, like dedicated in the commented code.

Like the code is right now, the above `ExampleContextKey` compiles as expected. The compiler warning wishes to refactor `ContextKey` instead to ...

```swift
protocol ContextKey: OptionalContextKey where Value == DefaultValue {
    associatedtype DefaultValue

    static var defaultValue: DefaultValue { get }
}
```

... which is seemingly not the same. First of all it is not the same as to my understanding how the `where` clause works, and indeed the `ExampleContentKey` doesn't compile anymore without explicitly adding the `typealias Value = String`.

If anyone has a better workaround for this, feel free to suggest a change 🎉

PSchmiedmayer replied:
@Supereg Thanks for the detailed explanation!

As you said an alternative to the `DefaultValue` and the extension below would be to just state 
```swift
    static var defaultValue: Self.Value { get }
```
As sated this would produce no compiler warning in this section as it is defined by the explicit declaration that `Value` is the associatedtype of `OptionalContextKey `.

As you stated the downside would be that you currently have to manually define an `typealias` like this to make the Swift compiler happy as it seems like it can't solve the type constraints across this protocol relationship at the moment:
```swift
struct OperationContextKey: ContextKey {
    typealias Value = Operation
    
    static var defaultValue: Operation = .automatic
}
```

II think this manual definition of a `typealias` would be an acceptable approach for me as this approach would reduce the number of `associatedtype`s by one and once the swift compiler gets smart enough this should be easier solvable then the current implementation with a second `associatedtype`. 
The two `associatedtype`s would point to the same type which might not be immediately visible when taking a look at the type definitions without taking a look at the externs below.

In the context of steadily reducing warnings on develop I would consider this a more likable solution as `ContextKey` and `OptionalContextKey` are both internal and not part of the public API, what do you think @Supereg @moritzschuell ?

moritzschuell replied:
@Supereg @PSchmiedmayer I also don't think that having to manually define the `typealias` would be a problem. And also, it somehow feels like the less "hacky" variant to me 🤓

bauer-andreas replied:
Feedback incorporated with ff2e55e

moritzschuell commented on Sources/Apodini/SyntaxTreeVisitor/ContextNode.swift:
```diff
@@ -28,31 +29,43 @@ class ContextNode {
+    func addContext<C: OptionalContextKey>(_ contextKey: C.Type = C.self, value: C.Value, scope: Scope) {
+        let newValue: C.Value
+
+        if C.Value.self is ExpressibleByNilLiteral.Type {
```
I am not sure whether this check is really the best way to determine if `Value` is an optional type.
There might be cases where custom types conform to `ExpressibleByNilLiteral`.

~~I search-engined a bit, and found that [`Mirror.DisplayStyle`](https://developer.apple.com/documentation/swift/mirror/displaystyle) has an `optional` case. 
Maybe that's a nicer approach, even though still only a runtime-check~~ (doesn't seem to work)

bauer-andreas replied:
> There might be cases where custom types conform to ExpressibleByNilLiteral.

Our stance is, we don't allow `nil` literal, and thus I think the check for that protocol is actually valid(?). We do not just blacklist `Optional` type, but every type which allows `nil`(?).

In any case, this is internal API, so not sure if that is that big of a problem. But I assume adding a private protocol and conformance for that protocol to `Optional` would also be valid.

PSchmiedmayer replied:
@Supereg @moritzschuell The `Runtime` framework offers a build in check to detect an optional type, we should be able to use this here. I have experimented a bit with this and created a test case to evaluate the function and it seems to work in my limits tests:
```swift
func testIsOptional() throws {
    func isOptional<T>(_ type: T.Type = T.self) throws -> Bool {
        let typeInfo = try Runtime.typeInfo(of: type)
        return typeInfo.kind == .optional
    }
    
    /// A custom type
    struct Test {}
    
    XCTAssertEqual(try isOptional(String.self), false)
    XCTAssertEqual(try isOptional(Int.self), false)
    XCTAssertEqual(try isOptional(Test.self), false)
    XCTAssertEqual(try isOptional(Optional<Test>.self), true)
    XCTAssertEqual(try isOptional(Optional<Test>.self), true)
    XCTAssertEqual(try isOptional(String?.self), true)
    XCTAssertEqual(try isOptional(String??.self), true)
    XCTAssertEqual(try isOptional(String???.self), true)
    XCTAssertEqual(try isOptional(Never.self), false)
}
```

Therefore we could introduce an internal method similar to `assertTypeIsStruct` that can look like this: 
```swift
func isOptional<T>(_ type: T.Type = T.self) throws -> Bool {
    let typeInfo = try Runtime.typeInfo(of: type)
    return typeInfo.kind == .optional
}
```
and could be used here:
```suggestion
        if isOptional(C.Value.self) {
```

bauer-andreas replied:
I incorporated your proposed change @PSchmiedmayer with 44b82fb.
I just modified the isOptional func to not be throwing (with documentation why this is possible).
Basically `typeInfo` only throws if the `kind` is something it does not support (or is not a "type"). If the kind is optional it will never throw, as it will generate a enum TypeInfo. Thus if typeInfo throws, we know that it can't be an optional.
I think that makes using the function a bit easier.

I also just added the test cases you provided 👍

PSchmiedmayer replied:
Thats a good point, making it non throwing is a good addition 👍

moritzschuell reviewed (commented):
Great stuff! Thanks for this PR.
I left a comment on the runtime-check for optional types, otherwise this looks good to me.

bauer-andreas committed:
Replace hacky associatedtypes thing with less hacky thing

bauer-andreas committed:
Add proper isOptional check

bauer-andreas commented:
Depending on what gets merged first, #117 could make use of this new `OptionalContextKey` protocol.

bauer-andreas committed:
Move RelationshipNameModifier to an OptionalContextKey

moritzschuell reviewed (approved):
Looks good to me 🙂 Thanks for all your detailed explanations!

PSchmiedmayer reviewed (approved):
Thanks for incorporating the changes! Looks like a great addition to Apodini 🎉

bauer-andreas commented:
Okay what do I doo with CodeCov falling with `77.64% (-0.00%) compared to 370e249` 😅 🙃

bauer-andreas commented:
administrator magic rules ✨

PSchmiedmayer commented:
The code coverage diff was so minor, I merged the branch into develop 👍
